3d datasets

petewa opened the pull request:
### Reason for this pull request

Extend ODC to support indexing and loading of 3D datasets, e.g. `(t, z, y, x)`

See Enhancement Proposal: https://github.com/opendatacube/datacube-core/wiki/ODC-EP-001---Add-Support-for-3D-Datasets

Code is backwards compatible and all existing tests pass.

### Proposed changes
   - Add support for indexing and loading 3D datasets for both `xarray` and `dask` load paths.
   - Code changes:
     - `datacube/api/core.py`
       - Changes to `_xr_load` and `_dask_load` to support 3D Datasets
       - `dc.load` API is unchanged and works seamlessly with the new extra dimension.
     - `datacube/api/query.py` and `datacube/index/hl.py`
       - Handle queries on extra dimensions in `dc.load`
     - `datacube/model/__init__.py`
       - Generate 2D measurements from a 3D measurement definition
       - `ExtraDimensions` class to represent the extra dimension with array helping methods to subset into a dimension.
     - `datacube/model/schema/dataset-type-schema.yaml`
       - New definition for the extra dimensionality
     - `datacube/storage/_load.py` and `datacube/storage/_read.py`
       - Handle reading from a 3D dataset
     - `examples/io_plugin/dcio_example/xarray_3d.py`
       - Example 3D driver based off the `pickle` and our `datacube-zarr` 3D driver.
     - `examples/io_plugin/setup.py`
       - Added the sample `xarray_3d` driver.
     - `tests/test_3d.py` and `integration_tests/test_3d.py`
       - Tests
     - `tests/test_utils_aws.py` and `tests/test_utils_dask.py`
       - Update tests to support new boto version.

 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

<!--
See https://github.com/blog/2111-issue-and-pull-request-templates for more
information on pull request templates.

-->

petewa committed:
Update tests for new version of boto

petewa committed:
First lot of changes to update ODC for 3D

petewa committed:
Remove debug statements

petewa committed:
Added spectral_definition_map to 3D measurement definition.
Minor housekeeping.

petewa committed:
Support single z value query in dc.load()
e.g. z=(5) or z=5.

petewa committed:
Handle dim_name=None query in dc.load() for 3D datasets.

petewa committed:
Fix issue in _xr_load and _make_dask_array where it would always retrieve values from indexed from 0 regarding of the requested dimension range.

rtaib committed:
Fixed incorrect variable name and some typos.

petewa committed:
Fix for loading 3D datasets with a reversed z query e.g. z=(20, 10).

petewa committed:
Fix issue with extra_dimensions with find_datasets when product is unknown.

rtaib committed:
3D datasets tests.

Require the xarray_3d driver and gedi data, both part of this commit.

rtaib committed:
Removed outdated code.

petewa committed:
Enable triggering Github Workflow checks on the examples directory, since tests now use the sample drivers.

petewa committed:
Added documentation

Kirill888 commented on datacube/model/__init__.py:
```diff
@@ -504,6 +535,86 @@ def extract_point(name):
+        """Validate extra_dimension metadata in the product definition.
+
+        :param definition: Dimension definition dict, typically retrieved from the product definition's
+            `extra_dimensions` field.
```
A brief description of what this is doing would be nice. What is this code checking, what failure modes it is supposed to detect.

petewa replied:
Will do.

petewa replied:
Done, Please let me know if there are any other required changes.

Kirill888 commented on tests/data/lbg/gedi/GEDI02_B_2019228092851_O03828_T04236_02_001_01/__GEDI02_B_2019228092851_O03828_T04236_02_001_01_gedi_l2b_cover_z.yaml:
```diff
@@ -0,0 +1,120 @@
+      - 0.0
+      - 1.0
+measurements:
+  "5":
+    layer: array
+    path: ./GEDI02_B_2019228092851_O03828_T04236_02_001_01_cover_z.xarray_3d
+  "10":
+    layer: array
+    path: ./GEDI02_B_2019228092851_O03828_T04236_02_001_01_cover_z.xarray_3d
+  "15":
+    layer: array
+    path: ./GEDI02_B_2019228092851_O03828_T04236_02_001_01_cover_z.xarray_3d
+  "20":
+    layer: array
+    path: ./GEDI02_B_2019228092851_O03828_T04236_02_001_01_cover_z.xarray_3d
+  "25":
+    layer: array
+    path: ./GEDI02_B_2019228092851_O03828_T04236_02_001_01_cover_z.xarray_3d
+  "30":
+    layer: array
+    path: ./GEDI02_B_2019228092851_O03828_T04236_02_001_01_cover_z.xarray_3d
+  "35":
+    layer: array
+    path: ./GEDI02_B_2019228092851_O03828_T04236_02_001_01_cover_z.xarray_3d
+  "40":
+    layer: array
```
I'm confused about this, I thought there would be only one measurement per ND dataset, and product definition would cover how to interpret extra dimension. But this looks like you have to list all, potentially hundreds, of sub-bands. We could that before.

petewa replied:
Will do.

Most of these were for backwards compatibility for loading a 3D measurement as though they are a list of 2D measurements. A 3D measurement is the main measurement, and a bunch of 2D measurement are a auto-generated from the 3D measurement. The 2D auto-generated components can go.

It means you can only do a
 - `dc.load(...measurements=['cover_z'], z=(5,10)...)`
    - Gives you a 3D result
 - and not a 
 - `dc.load(...measurements=['5', '10']')`
    - Give you a list of 2D results.

I'll remove the 2D auto-generated stuff as well as making the dataset yaml a single 3D measurement with one uri.

petewa replied:
Done, Please let me know if there are any other required changes.

petewa committed:
Address PR (remove auto-generated 2D measurements, measurement_map and alias_map)
spectral_definition now supports a list of definitions.
Added --allow-multiple-documents to .pre-commit-config.yaml to allow multi-document definitions
Docs to follow

dtpc committed:
Documentation and examples for 3D datasets

petewa committed:
Added comments to validate_extra_dims.
Updated whats_new.rst with PR number.

Kirill888 commented on datacube/model/__init__.py:
```diff
@@ -460,8 +463,27 @@ def fix_nodata(m):
-                                                       for m in self.definition.get('measurements', []))
+            self._canonical_measurements = OrderedDict()
+            for m in self.definition.get('measurements', []):
+                # Generate 2D measurements from a 3D measurement so we dont have to fully enumerate them
+                # in the product definition
+                # Note: Might be a good idea to natively support 3D measurements
+                # Todo: See if we can make a sister change to eo3 to describe a 3D datasource
+                '''
+                if 'extra_dim' in m:
+                    for idx, mm in enumerate(m['extra_dim'].get('measurement_map', [])):
+                        new_m = deepcopy(m)
+                        new_m.update({'name': mm})
+                        new_m.update({'extra_dim_index': idx})
+                        if 'alias_map' in m['extra_dim']:
+                            new_m.update({'alias': m['extra_dim']['alias_map'][idx]})
+                        if 'spectral_definition_map' in m['extra_dim']:
+                            new_m.update({'spectral_definition': m['extra_dim']['spectral_definition_map'][idx]})
+                        new_m.pop('extra_dim', None)
+                        self._canonical_measurements.update({new_m['name']: Measurement(**fix_nodata(new_m))})
+                '''
```
Let's not keep commented out code.

petewa replied:
Will do.

petewa replied:
Actually this can go back to the original code in develop.

petewa replied:
Done, Please let me know if there are any other required changes.

Kirill888 commented on datacube/model/__init__.py:
```diff
@@ -840,3 +927,183 @@ def metadata_from_doc(doc: Mapping[str, Any]) -> MetadataType:
+    def __getitem__(self, dim_slices: Dict[str, Union[float, Tuple[float, float]]]) -> "ExtraDimensions":
+        """Return a ExtraDimensions subsetted by dim_slices
+
+        :param dim_slices: Optional cache to re-use geoboxes instead of creating new one each time
```
is this document comment valid or a cut and paste?

petewa replied:
Was a cut and paste, will fix this.

petewa replied:
Done, Please let me know if there are any other required changes.

Kirill888 commented on datacube/model/__init__.py:
```diff
@@ -840,3 +927,183 @@ def metadata_from_doc(doc: Mapping[str, Any]) -> MetadataType:
+            # Adjust slices relative to original.
+            if dim_name in ed._dim_slice:
+                ed._dim_slice[dim_name] = (
+                    ed._dim_slice[dim_name][0] + dim_slice[0],
+                    ed._dim_slice[dim_name][0] + dim_slice[1],
```
Not sure what this does, but are we sure it's not meant to be this:

```suggestion
                    ed._dim_slice[dim_name][0] + dim_slice[0],
                    ed._dim_slice[dim_name][1] + dim_slice[1],
```

petewa replied:
This should be correct, since there's only 1 dimension, there's only 1 origin / point of reference.

petewa replied:
Done, Please let me know if there are any other required changes.

petewa committed:
Remove commented block - revert  back to orginal code measurements.
Replace cut and paste comment in ExtraDimensions.__getitem__ with a proper one.

omad reviewed (approved):
Thanks Peter

RichardScottOZ commented:
@petewa On reading, this is then designed for when you know the other dimensions in advance... e.g. discrete depth slices etc?

SpacemanPaul commented:
> @petewa On reading, this is then designed for when you know the other dimensions in advance... e.g. discrete depth slices etc?

@RichardScottOZ 

Yes.  Because the dimension is modeled as a collection of measurements within a dataset.  

To support arbitrary dimensions you would need to model each dimension slice as a separate dataset (as we do with time for example).  That's a valid approach, but it's a whole other PR.

woodcockr commented:
The design approach in the existing implementation was made on the basis of providing a minimal change impact on the existing code base. The preference would of course be the "other PR" as @SpacemanPaul put it.  CSIRO will be revisiting this in the coming year as we need more advanced hyperspectral support (amongst other things) where the back end dataset  change is required (along with the more arbitrary already Zarr driver plugin in the other repo that goes with this and the refactoring on the database and storage mechanisms @SpacemanPaul  is chipping away at).
